### PR TITLE
[BEAM-2855] nexmark python suite implement queries 0, 1, 2 and 9

### PR DESCRIPTION
**Pull request by leiyiz**
changed queries 0-2, implemented query9, coders for all models, new models for query results, changed launcher to count and print results to file

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by leiyiz** [comment]
R: @pabloem 
R: @y1chi

**Comment by y1chi** [comment]
Could you rename the title to be more meaningful and add `[BEAM-2855]` in the front so that it will be hooked with the right JIRA ticket https://issues.apache.org/jira/browse/BEAM-2855.

File: sdks/python/apache_beam/testing/benchmarks/nexmark/models/nexmark_model.py [hunk 1]
```diff
@@ -26,31 +26,55 @@
...
   def __repr__(self):
-    return 'Person({id}, {email})'.format(
-        **{
-            'id': self.id, 'email': self.email
-        })
+    return nexmark_util.model_to_json(self)
+
+
+class AuctionCoder(FastCoder):
+  def to_type_hint(self):
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/models/nexmark_model.py:64]: why we need to override and pass here?

File: sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_launcher.py [hunk 2]
```diff
@@ -263,4 +279,4 @@ def run(self):
 if __name__ == '__main__':
   launcher = NexmarkLauncher()
   launcher.run()
-  launcher.cleanup()
+  # launcher.cleanup()
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_launcher.py:282]: I think this shouldn't be commented out.

File: sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py [hunk 3]
```diff
@@ -103,6 +121,108 @@ def process(self, elem):
...
+      cand, (nexmark_model.Auction, nexmark_model.Bid, nexmark_model.Person)):
+    return construct_json_dict(cand)
+  else:
+    return cand
+
+
+def millis_to_timestamp(millis: int):
+  second = millis // 1000
+  micro_second = millis % 1000 * 1000
+  return Timestamp(second, micro_second)
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py:228]: I think you can simply pass the micro converted from millis to Timestamp

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    self.is_auction_window = is_auction_window
+
+  @staticmethod
+  def for_auction(timestamp, auction: nexmark_model.Auction):
+    return AuctionOrBidWindow(timestamp, auction.expires, auction.id, True)
+
+  @staticmethod
+  def for_bid(expected_duration_micro, timestamp, bid: nexmark_model.Bid):
+    return AuctionOrBidWindow(
+        timestamp, timestamp + expected_duration_micro * 2, bid.auction, False)
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py:44]: I think you need to be careful here, if timestamp is Timestamp type and expected_duration_micro is int, `+` will try to convert expected_duration_micro * 2 with Duration.of(expected_duration_micro * 2), which by default treat the expected_duration_micro * 2 as seconds value. You probably want to do something like timestamp + Duration(micros=expected_duration_micro * 2).

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+          AuctionOrBidWindow.for_bid(
+              self.expected_duration, assign_context.timestamp, event)
+      ]
+    else:
+      raise ValueError(
+          '%s can only assign windows to auctions and bids, but received %s' %
+          (self.__class__.__name__, event))
+
+  def merge(self, merge_context):
+    id_to_auction = {}
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py:102]: nit: rename to auction_id_to_auction_window, auction_id_to_bid_window.

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    id_to_bid = {}
+    for window in merge_context.windows:
+      if window.is_auction_window_fn():
+        id_to_auction[window.auction] = window
+      else:
+        if window.auction in id_to_bid:
+          bid_windows = id_to_bid[window.auction]
+        else:
+          bid_windows = []
+          id_to_bid[window.auction] = bid_windows
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py:112]: nit: you can do
```
if window.auction not in id_to_bid:
  id_to_bid[window.auction] = []
id_to_bid[window.auction].append(window)
```

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    if auction is None:
+      return
+    best_bid = None
+    for bid in group[nexmark_query_util.BID_TAG]:
+      if bid.price < auction.reserve:
+        continue
+      if best_bid is None or JoinAuctionBidFn.higher_bid(bid, best_bid):
+        best_bid = bid
+    if best_bid is None:
+      return
```
Comment (y1chi) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py:156]:
```
if best_bid:
  yield ...
```

File: sdks/python/apache_beam/testing/benchmarks/nexmark/models/nexmark_model.py (hunk 1 above)
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/models/nexmark_model.py:64]: my bad it is a mistake..

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  @staticmethod
+  def for_auction(timestamp, auction: nexmark_model.Auction):
+    return AuctionOrBidWindow(timestamp, auction.expires, auction.id, True)
+
+  @staticmethod
+  def for_bid(expected_duration_micro, timestamp, bid: nexmark_model.Bid):
+    return AuctionOrBidWindow(
+        timestamp, timestamp + expected_duration_micro * 2, bid.auction, False)
+
+  def is_auction_window_fn(self):
```
[the remark this answers is not in the document]
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/queries/winning_bids.py:46]: I think it is fine to directly access the is_auciton_window field

**Comment by y1chi** [comment]
R: @pabloem

File: sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py [hunk 9]
```diff
@@ -103,6 +121,107 @@ def process(self, elem):
...
+    person: {id,name,email,credit_card,city, \
+                          state,timestamp,extra}
+    auction: {id,item_name, description,initial_bid, \
+                          reserve_price,timestamp,expires,seller,category,extra}
+    bid: {auction,bidder,price,timestamp,extra}
+
+  For example:
+
+    {"id":1000,"name":"Peter Jones","emailAddress":"nhd@xcat.com",\
+        "creditCard":"7241 7320 9143 4888","city":"Portland","state":"WY",\
```
Comment (pabloem) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py:138]: can you improve the formatting of these docs?

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query2.py [hunk 10]
```diff
@@ -29,17 +29,15 @@
...
 import apache_beam as beam
-from apache_beam.testing.benchmarks.nexmark.models import nexmark_model
-from apache_beam.testing.benchmarks.nexmark.nexmark_util import ParseEventFn
-from apache_beam.testing.benchmarks.nexmark.nexmark_util import display
+from apache_beam.testing.benchmarks.nexmark.models import auction_price
+from apache_beam.testing.benchmarks.nexmark.queries import nexmark_query_util
 
 
-def load(raw_events, metadata=None):
+def load(events, metadata=None):
```
[the remark this answers is not in the document]
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query2.py:36]: yeah that is probably better

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query1.py [hunk 11]
```diff
@@ -29,20 +29,16 @@
...
-      beam.Filter(lambda event: isinstance(event, nexmark_model.Bid))
+      | nexmark_query_util.JustBids()
       | 'ConvertToEuro' >> beam.Map(
           lambda bid: nexmark_model.Bid(
               bid.auction,
-              bid.bidder, (float(bid.price) * 89) // 100,
-              bid.timestamp,
-              bid.extra))
-      | 'DisplayQuery1' >> beam.Map(display))  # pylint: disable=expression-not-assigned
+              bid.bidder, (bid.price * 89) // 100,
```
[the remark this answers is not in the document]
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query1.py:42]: no problem

File: sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query2.py [hunk 12]
```diff
@@ -29,17 +29,15 @@
...
   return (
-      raw_events
-      | 'ParseEventFn' >> beam.ParDo(ParseEventFn())
-      | 'FilterInAuctionsWithSelectedId' >> beam.Filter(
-          lambda event: (
-              isinstance(event, nexmark_model.Auction) and event.id == metadata.
-              get('auction_id')))
-      | 'DisplayQuery2' >> beam.Map(display))  # pylint: disable=expression-not-assigned
+      events
+      | nexmark_query_util.JustBids()
```
[the remark this answers is not in the document]
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query2.py:39]: yes the doc should be updated

File: sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py (hunk 9 above)
Comment (leiyiz) [reply sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py:138]: I'm not exactly sure what format should this be in? should I just delete these examples or something

File: sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py [hunk 13]
```diff
@@ -103,6 +121,107 @@ def process(self, elem):
...
+        "dateTime":1528098831026,\"extra":"WN_HS_bnpVQ\\[["}
+
+    {"id":1000,"itemName":"wkx mgee","description":"eszpqxtdxrvwmmywkmogoahf",\
+        "initialBid":28873,"reserve":29448,"dateTime":1528098831036,\
+        "expires":1528098840451,"seller":1000,"category":13,"extra":"zcuupiz"}
+
+    {"auction":1000,"bidder":1001,"price":32530001,"dateTime":1528098831066,\
+        "extra":"fdiysaV^]NLVsbolvyqwgticfdrwdyiyofWPYTOuwogvszlxjrcNOORM"}
```
Comment (pabloem) [inline sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_util.py:146]: It's a bit of a silly thing, but something like this:

```suggestion
    person: {id,name,email,credit_card,city, \
             state,timestamp,extra}
    auction: {id,item_name, description,initial_bid, \
              reserve_price,timestamp,expires,seller,category,extra}
    bid: {auction,bidder,price,timestamp,extra}

  For example:

    {"id":1000,"name":"Peter Jones","emailAddress":"nhd@xcat.com",\
     "creditCard":"7241 7320 9143 4888","city":"Portland","state":"WY",\
     "dateTime":1528098831026,\"extra":"WN_HS_bnpVQ\\[["}

    {"id":1000,"itemName":"wkx mgee","description":"eszpqxtdxrvwmmywkmogoahf",\
     "initialBid":28873,"reserve":29448,"dateTime":1528098831036,\
     "expires":1528098840451,"seller":1000,"category":13,"extra":"zcuupiz"}

    {"auction":1000,"bidder":1001,"price":32530001,"dateTime":1528098831066,\
     "extra":"fdiysaV^]NLVsbolvyqwgticfdrwdyiyofWPYTOuwogvszlxjrcNOORM"}
```

**Comment by pabloem** [comment]
: )

**Pull request merged**